scipy/stats/stats.py

```diff
@@ -3294,26 +3294,29 @@ def spearmanr(a, b=None, axis=0, nan_policy='propagate'):
     """
     a, axisout = _chk_asarray(a, axis)
 
-    contains_nan, nan_policy = _contains_nan(a, nan_policy)
+    a_contains_nan, nan_policy = _contains_nan(a, nan_policy)
 
-    if contains_nan and nan_policy == 'omit':
+    if a_contains_nan and nan_policy == 'omit':
         a = ma.masked_invalid(a)
-        b = ma.masked_invalid(b)
-        return mstats_basic.spearmanr(a, b, axis)
 
-    if a.size <= 1:
+    if a.size <= 1 or (a_contains_nan and nan_policy == 'propagate'):
         return SpearmanrResult(np.nan, np.nan)
+        
     ar = np.apply_along_axis(rankdata, axisout, a)
 
     br = None
     if b is not None:
         b, axisout = _chk_asarray(b, axis)
 
-        contains_nan, nan_policy = _contains_nan(b, nan_policy)
+        b_contains_nan, nan_policy = _contains_nan(b, nan_policy)
+
+        if a_contains_nan or b_contains_nan:
+            if nan_policy == 'propagate':
+                return SpearmanrResult(np.nan, np.nan)
```

Comment:
Hmmm. I think it should preserve the output shape for n-dim inputs

```
In [20]: corr, pvals = spearmanr([[1, 2, 3], [5, 6, 7]], [[5, 6, 11], [8, 9, np.nan]], nan_policy='propagate')

In [21]: pvals
Out[21]: nan
```

while with scipy 0.18.1,

```
In [5]:  corr, pvals = spearmanr([[1, 2, 3], [5, 6, 7]], [[5, 6, 11], [8, 9, np.nan]], nan_policy='propagate')

In [6]: pvals
Out[6]: 
array([[ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan],
       [ nan,  nan,  nan,  nan,  nan,  nan]])
```